File: 1147-flip-columns-for-maximum-number-of-equal-rows/flip-columns-for-maximum-number-of-equal-rows.ex
defmodule Solution do
  @spec max_equal_rows_after_flips(matrix :: [[integer]]) :: integer
  def max_equal_rows_after_flips(matrix) do
     # Map to store the frequency of patterns
    bit_map =
      Enum.reduce(matrix, %{}, fn row, acc ->
        # Construct the pattern based on the first element of the row
        pattern =
          if hd(row) == 0 do
            Enum.map(row, &Integer.to_string(&1))
          else
            Enum.map(row, &Integer.to_string(1 - &1))
          end
          |> Enum.join("")

        # Update the frequency map
        Map.update(acc, pattern, 1, &(&1 + 1))
      end)

    # Find the maximum value in the frequency map
    bit_map
    |> Map.values()
    |> Enum.max()
  end
end

File: 1147-flip-columns-for-maximum-number-of-equal-rows/flip-columns-for-maximum-number-of-equal-rows.go
func maxEqualRowsAfterFlips(mat [][]int) int {
    patFreq := make(map[string]int)
    
    for _, row := range mat {
        pattern := make([]byte, len(row))
        if row[0] == 0 {
            for i, bit := range row {
                pattern[i] = byte(bit) + '0'
            }
        } else {
            for i, bit := range row {
                pattern[i] = byte(bit ^ 1) + '0'
            }
        }
        patFreq[string(pattern)]++
    }
    
    maxFreq := 0
    for _, freq := range patFreq {
        if freq > maxFreq {
            maxFreq = freq
        }
    }
    return maxFreq
}